src/hotspot/os/aix/os_aix.cpp

```diff
@@ -1167,23 +1167,28 @@ static void* dll_load_library(const char *filename, char *ebuf, int ebuflen) {
 */
 void *os::dll_load(const char *filename, char *ebuf, int ebuflen) {
   void* result = nullptr;
-  unsigned long buffer_length = strlen(filename);
-  int extension_length = 3;
-  char* file_path = NEW_C_HEAP_ARRAY(char, buffer_length + extension_length + 1, mtInternal);
-  strncpy(file_path,filename, buffer_length + 1);
+  char *file_path = strdup(filename);
   char* const pointer_to_dot = strrchr(file_path, '.');
+  char const *old_extension = ".so";
+  char const *new_extension = ".a";
   if (pointer_to_dot == nullptr) {
     log_info(os)("Attempting to load a shared object without extension %s", filename);
     FREE_C_HEAP_ARRAY(char, file_path);
     return result;
   }
   // First try to load the existing file.
-  result = dll_load_library(file_path, ebuf, ebuflen);
+  result = dll_load_library(filename, ebuf, ebuflen);
+  int error_code = errno;
```

Comment:
this might not necessarily be the `errno` of the underlying `dlopen()`, because there is to much code in-between and branches without a `dlopen()` call.

Reply:
Suchi, errno is a global static variable. If some runtime API sets it, it will continue to have this value until the next runtime call updates it. If you call dll_load_library, there are many execution paths not passing dlopen(). So you receive an errno from some unknown runtime API called before. The correct errno handling is:
```
errno = 0;
runtime_API_which_might_set_errno_in_error_case();
error_code = errno;
```
But what you really need is the result of the `search_file_in_LIBPATH(...)` call in `Aix_dlopen()`. If it is false, then the error_report starts with the string "Could not load module . ....."
This is called in any case. A `dlopen()` is not called in any case.

Reply:
Thanks for the detailed explanation @JoKern65 . Do then in this errno check may not be necessary ? or can we still set errno and access it some way ?

Reply:
In this special case here I would not use errno, but the string returned in ebuf, in case the result is nullptr.

Reply:
Argh, I keep forgetting `dlopen` does not set errno (at least its not guaranteed by the standard).

Ok, in this case, I would not analyze the string either because it may be locale-dependent.

Never mind then; I'd thought this would be easy. One could probably use load() instead of dlopen() on AIX, but that change would require more investigations to see if load() and dlopen() are equivalent.

Okay, in that case just skip the errno test and call the second dlopen directly. Lets hope for the best then.

Reply:
Sure .thank for all the info. This is great learning.

Reply:
reverted the error code handling.